server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,186 @@
+func MappingsFromJSON(data []byte) (Mappings, error) {
+	var mappings Mappings
+	if err := json.Unmarshal(data, &mappings); err != nil {
+		return Mappings{}, err
+	}
+	return mappings, nil
+}
+
+// ToJSON serializes a list of app mappings to json.
+func (m *Mappings) ToJSON() []byte {
+	b, _ := json.Marshal(m)
+	return b
+}
+
+// SynchronizeApps synchronizes apps upgrading, downgrading and deleting apps
+func (adm *Admin) SynchronizeApps() error {
+	mappingsFile := os.Getenv(appsMappingsEnvVarName)
+	if mappingsFile == "" {
+		return nil
+	}
+	list := strings.Split(mappingsFile, "/")
+	if len(list) != 2 {
+		return errors.Errorf("Wrong format of an env var - %s", mappingsFile)
+	}
+	bucket := list[0]
+	key := list[1]
+
+	resp, err := adm.awsClient.S3FileDownload(bucket, key)
+	if err != nil {
+		return errors.Wrapf(err, "can't download file %s/%s", bucket, key)
+	}
+
+	mappings, err := MappingsFromJSON(resp)
+	if err != nil {
+		return errors.Wrapf(err, "can't deserialize mappings file %s/%s", bucket, key)
+	}
+
+	registeredApps := adm.store.ListApps()
+
+	registeredAppsMap := map[api.AppID]*api.App{}
+	updatedAppVersionsMap := map[api.AppID]string{}
+	// Update apps
+	for _, registeredApp := range registeredApps {
+		registeredAppsMap[registeredApp.Manifest.AppID] = registeredApp
+		manifest, ok := mappings.Apps[registeredApp.Manifest.AppID]
+		if !ok {
+			// TODO should deleting the app be that easy?
+			if err := adm.DeleteApp(registeredApp); err != nil {
+				return errors.Wrapf(err, "can't delete an app")
+			}
+		}
+		if manifest.Version != registeredApp.Manifest.Version {
+			// update app in proxy plugin
+			oldVersion := registeredApp.Manifest.Version
+			registeredApp.Manifest.Version = manifest.Version
+			if err := adm.store.StoreApp(registeredApp); err != nil {
+				return errors.Wrapf(err, "can't store app - %s", registeredApp.Manifest.AppID)
+			}
+			updatedAppVersionsMap[registeredApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as registered
+	for appID, manifest := range mappings.Apps {
+		if _, ok := registeredAppsMap[appID]; !ok {
+			if err := adm.AddApp(manifest); err != nil {
+				return errors.Wrapf(err, "can't add new app as registered")
+			}
+		}
+	}
+
+	registeredAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
+	for _, registeredApp := range registeredAppsUpgraded {
+		if registeredApp.Status == api.AppStatusEnabled {
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[registeredApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[registeredApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app
+			if err := adm.call(registeredApp, registeredApp.Manifest.OnStartup, values); err != nil {
+				adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *api.App) error {
```

Comment:
2/5 We need to delete the bot account also.

Reply:
Agree, but there is not endpoint for deleting the bots. I'll add a ticket for that